### PR TITLE
Mandatory update status reporting

### PR DESCRIPTION
**Pull request by bboozzoo**
@kacf @pasinskim

### CODE REVIEW

File: state.go [hunk 1]
```diff
@@ -584,19 +585,75 @@ func NewUpdateErrorState(err menderError, update UpdateResponse) State {
...
+
+func (usr *UpdateStatusReportState) Handle(ctx *StateContext, c Controller) (State, bool) {
+	if err := StoreStateData(ctx.store, StateData{
+		Id:           usr.Id(),
+		UpdateInfo:   usr.update,
+		UpdateStatus: usr.status,
+	}); err != nil {
+		log.Errorf("failed to store state data in update status report state: %v",
+			err)
+		// TODO: what now?
```
Comment (kacf) [inline state.go]: Should be the same as failure to report I think, rollback and reboot.

File: state.go [hunk 2]
```diff
@@ -584,19 +585,75 @@ func NewUpdateErrorState(err menderError, update UpdateResponse) State {
...
+						Timestamp: time.Now().Format(time.RFC3339),
+						Level:     "error",
+						Message:   "update failed",
+					},
+				})
+			}
+		}
+
+		log.Debug("reporting complete")
+		break
```
Comment (kacf) [inline state.go]: Won't this `break` always be hit? It looks like you're trying to do some repeated attempts to upload the log, but I cannot see that the loop can ever be executed more than once, not even with the currently commented code.

File: state_test.go [hunk 3]
```diff
@@ -195,13 +198,65 @@ func TestStateUpdateError(t *testing.T) {
 		store: ms,
 	}
 	sc := &stateTestController{}
-	es = NewUpdateErrorState(fooerr, UpdateResponse{})
-	es.Handle(&ctx, sc)
+	es = NewUpdateErrorState(fooerr, update)
+	s, _ := es.Handle(&ctx, sc)
+	assert.IsType(t, &UpdateStatusReportState{}, s)
+	// verify that update status report state data is correct
+	usr, _ := s.(*UpdateStatusReportState)
```
Comment (kacf) [inline state_test.go]: Minor: Maybe write without `_` to catch the error that it is the wrong type?

Comment (bboozzoo) [inline state_test.go]: There's an assert just above, `assert.IsType(t, &UpdateStatusReportState{}, s)`

File: state.go (hunk 1 above)
Comment (bboozzoo) [inline state.go]: Makes sense. Since we don't support rollback at the moment, I'll update the comment to reflect that.

File: state_test.go (hunk 3 above)
Comment (kacf) [inline state_test.go]: Right, doh!

File: state.go [hunk 4]
```diff
@@ -573,19 +585,75 @@ func NewUpdateErrorState(err menderError, update UpdateResponse) State {
...
-		LogEntry{
-			Timestamp: time.Now().Format(time.RFC3339),
-			Level:     "error",
-			Message:   "update failed",
-		},
-	})
+	return NewUpdateStatusReportState(ue.update, statusFailure), false
+}
+
+// Wrapper for mandatory update state reporting, the state handler will attempt
```
Comment (pasinskim) [inline state.go]: ` Wrapper for mandatory update status reporting. The state handler will attempt`

**Comment by pasinskim** [comment]
@bboozzoo overall looks OK. Just fix minor issues if there are any outstanding. What is more, make sure to comment out the appropriate code so that client won't break if we can not send logs and status to the server. 
You can add in comments reference to https://tracker.mender.io/browse/MEN-536 (sending status) and possibly https://tracker.mender.io/browse/MEN-437 (sending logs).

**Pull request merged**